### PR TITLE
Update kokkos for new 3.x build systems

### PR DESCRIPTION
**Pull request by jjwilke**
* Move old kokkos to kokkos-legacy
* Move old kokkos-kernels to kokkos-kernels-legacy
* Add Kokkos and Kokkos-Kernels 3.0 and 3.1
* Add an nvcc_wrapper package

### CODE REVIEW

File: var/spack/repos/builtin/packages/kokkos-kernels-legacy/package.py
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/kokkos-kernels-legacy/package.py]: Could you re-order these `depends_on` to match the order of the versions above?

**Comment by adamjstewart** [comment]
This PR relates to the discussion in #15420.

File: var/spack/repos/builtin/packages/kokkos-kernels-legacy/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos-kernels-legacy/package.py]: I'd prefer to not modify the legacy package any more than just moving it?

Comment (junghans) [reply var/spack/repos/builtin/packages/kokkos-kernels-legacy/package.py]: Yeah, let's not touch it.

**Review by junghans — CHANGES_REQUESTED** [review]
I think we need to do a `s/kokkos/&-legacy/` on all spackages, too.

**Comment by jjwilke** [comment]
@junghans Cleaned up everything pointing to kokkos-legacy. Also removed all git-based versions from kokkos-legacy.

**Review by junghans — APPROVED** [review]

**Comment by crtrott** [comment]
Someone should check whether lammps works this way? I think they may need Kokkos 3.1 now.

**Comment by junghans** [comment]
> Someone should check whether lammps works this way? I think they may need Kokkos 3.1 now.

The lammps spackage is wrong, even though it depends on kokkos it uses it's bundled version. Will fix that in a follow-up PR.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (aprokop) [inline var/spack/repos/builtin/packages/kokkos/package.py:29]: What about `hip`?

**Comment by alalazo** [comment]
@junghans It might take a while before I can review this PR thoroughly, so feel free to proceed without my review in case it's urgent.

**Comment by junghans** [comment]
@adamjstewart please review.

**Comment by adamjstewart** [comment]
I think everything looks fine, although I would love to reach a consensus on #15420 before we change too many of these packages.

**Comment by junghans** [comment]
I don't see a big problem here, `lammps`, `arbox` and `cabana` can already use kokkos-3, so this PR would be a step forward. 

`examinimd` (cc @stanmoore1) and `miniaero` (@crtrott) will have to be updated.

Will #15420 allow to depend on `kokkos::legacy`?

**Comment by stanmoore1** [comment]
ExaMiniMD has been updated to use Kokkos 3.

**Comment by stanmoore1** [comment]
@junghans Not sure what else is needed for ExaMiniMD?

**Comment by adamjstewart** [comment]
> Will #15420 allow to depend on `kokkos::legacy`?

Yes, the syntax would be `depends_on('legacy.kokkos')`. The idea is that a package can depend on `kokkos`, and if you don't specify whether you want `builtin.kokkos` or `legacy.kokkos` then you can choose either one.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py:29]: Yeah, I should add that and remove ROCm. I need to figure out how to make variants version-dependent.

**Comment by jjwilke** [comment]
I suspect that no packages will depend on kokkos-legacy or kokkos-kernels-legacy very soon so the legacy thing may not be a big deal. The PR #15420 is way too long to read... I'm happy to make any changes, though, if there is an example package.py.

**Comment by junghans** [comment]
Well, we need to keep `kokkos-legacy` around for older versions of the packages.

**Comment by adamjstewart** [comment]
You may also be interested in #15896, which proposes a procedure for deprecating legacy software and eventually removing it from Spack. That may make a lot of the debate about a legacy repo less important.

**Comment by junghans** [comment]
> @junghans Not sure what else is needed for ExaMiniMD?

@stanmoore1, do you know about the state of `miniaero`?

**Comment by stanmoore1** [comment]
>@stanmoore1, do you know about the state of miniaero?

No, I've never run that.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (aprokop) [reply var/spack/repos/builtin/packages/kokkos/package.py:29]: As a workaround, you can just conflict.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (AndrewGaspar) [inline var/spack/repos/builtin/packages/kokkos/package.py]: It seems to me that having "host" arch variants is redundant with the arch provided by Spack, and I think Spack takes on the responsibility of being able to cross compile to a different backend arch from the front end. In fact, it looks like down on line 252 this is done, but then the variants are also applied.

Maybe having all these variants is the right way to do it, but I would like to see some discussion on whether the host variants can be eliminated in favor of just mapping spack arch's to the correct kokkos settings.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: There are a couple reasons. 
1. We may do more with the arch information than just add flags, in which case I need to give Kokkos the expected arch name
2. Not trusting Spack to always do the right thing

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py:29]: hip is added, along with missing vega architectures. I added the conflict for versions where appropriate.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (AndrewGaspar) [reply var/spack/repos/builtin/packages/kokkos/package.py]: For 1) that's fine - I propose mapping spack arch to `KOKKOS_ARCH_*`.

For 2), I guess I'd be curious if there are specific bugs... Maybe they could be fixed.

Comment (junghans) [reply var/spack/repos/builtin/packages/kokkos/package.py]: I think having variants and spack arch will just be confusing for the user. I would map the spack arch as well.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: 1) is already done, I think? Or did you mean something else? 
````
        kokkos_microarch_name = self.spack_micro_arch_map[spec.target.name]
        if kokkos_microarch_name:
            spack_microarches.append(kokkos_microarch_name)
        for arch in spack_microarches:
            options.append("-DKokkos_ARCH_%s=ON" % arch.upper())
````
I've been very conflicted on all of this. I don't want to add a kokkos-specific variant that already exists "natively" in Spack.  At the same time, I don't expect Kokkos and Spack to always have the exact same list of arches.

Comment (AndrewGaspar) [reply var/spack/repos/builtin/packages/kokkos/package.py]: That's what I was trying to say, perhaps poorly, in my comment. (Roughly) I want you to keep that bit, but delete this line for host-archs:
```
        self.append_args("ARCH",   self.arch_values, options)
```

I know that spack cares a lot of about capturing the breadth of micro-archs on HPC systems 100% right, and I've seen a lot of evidence of this - maybe @adamjstewart can comment on the work on this front.

**Comment by jjwilke** [comment]
Sounds like most approve and this is ready to go after the most recent changes?

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (junghans) [reply var/spack/repos/builtin/packages/kokkos/package.py]: ok, I see it now, it is in `spack_micro_arch_map`.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: So I went back and looked at our architecture list. The following are not covered by Spack as either a CUDA or Spack microarchitecture.
````
        'epyc': [False, 'Optimize for the EPYC architecture'],
        'fiji': [False, 'Optimize for the FIJI architecture'],
        'gfx901': [False, 'Optimize for the GFX901 architecture'],
        'vega900': [False, 'Optimize for the VEGA900 architecture'],
        'vega906': [False, 'Optimize for the VEGA906 architecture'],
````
I'm not sure why we use Ryzen and Epyc as opposed to Zen and Zen2 - so potentially we could move Epyc off this list. 

The rest are AMD GPU architectures, which I don't believe is covered by Spack at all, unless someone knows better.  I guess I could make a`gpu_arch` variant that takes those values. 

That would leave microarchitecture (target=), cuda_arch, and gpu_arch as the way to set architecture options. Does anyone have an opinion on this?

File: var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py
Comment (aprokop) [inline var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py]: Why does this point to a developer page? Shouldn't this be part of kokkos organization?

File: var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py
Comment (aprokop) [inline var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py]: Same as above.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (aprokop) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Probably don't need `+wrapper` here.

File: var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py]: Yeah, I'm just going to overrule others here. The nvcc_wrapper in kokkos was deprecated, but I brought it back. Fixed.

File: var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos-nvcc-wrapper/package.py]: Fixed. See above.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Oops. Fixed. Thank you.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Okay, I went ahead and just made an amd_gpu_arch option. It's sort of pointless, anyway, since there's no hipcc support in Spack yet.  Everything now uses Spack architecture options where possible.

**Review by aprokop — APPROVED** [review]
The parts I have some knowledge about looked good to me.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (aprokop) [reply var/spack/repos/builtin/packages/kokkos/package.py]: still there, it seems

Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Not for the last time on our VPN, the git push didn't go through. Nooowwww it's fixed.

**Review by AndrewGaspar — APPROVED** [review]

**Review by junghans — APPROVED** [review]

**Comment by junghans** [comment]
Let's see what breaks.....

**Comment by junghans** [comment]
As expected .... https://github.com/spack/spack/pull/16388

**Comment by aprokop** [comment]
@jjwilke Have you considered providing a default cuda arch, like `35`? Doesn't Kokkos use one by default when one has `Kokkos_ENABLE_CUDA=ON` but does not specify architecture? This would allow an easier basic installation of packages depending on `kokkos+cuda` to try them out. Otherwise, one needs to let users do more convoluted `spack install arborx ^kokkos cuda_arch=<number>`.

**Comment by jjwilke** [comment]
That thought has occurred to me. Last I checked, if you build with 35 on a 70, Kokkos crashes and says it got compute capability 0. You also wouldn't get optimized code. I'd probably prefer that Spack crashes and tells you what to change rather than go through all the way building Kokkos only to get a runtime error or silently get suboptimal performance. Not sure what the best tradeoff here is.

**Comment by AndrewGaspar** [comment]
It would be _possible_ to do machine reconnaissance and determine a good default for cuda similarly to how Spack sets the default arch.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (aprokop) [inline var/spack/repos/builtin/packages/kokkos/package.py:168]: Speaking of other things I noticed too late (i.e., after merge). What's the point of `wrapper` variant? Shouldn't  this simply be
```
depends_on("kokkos-nvcc-wrapper", when="+cuda")
```

I encountered this when working to update arborx (which has
`depends_on('kokkos+cuda', when='+cuda')`:
```
$ spack install arborx@0.9-beta+cuda ^kokkos cuda_arch=61
 -   ti623cj  arborx@0.9-beta%gcc@7.4.0 build_type=RelWithDebInfo +cuda~hip+mpi~openmp+serial arch=linux-ubuntu18.04-broadwell   
 -   h5te24i      ^kokkos@3.1%gcc@7.4.0~aggressive_vectorization amd_gpu_arch=none build_type=RelWithDebInfo ~compiler_warnings+cuda cuda_arch=70 +cuda_lambda~cuda_ldg_intrinsic~cuda_relocatable_device_code~cuda_uvm~debug~debug_bounds_check~debug_dualview_modify_check~deprecated_code~examples~explicit_instantiation~hip~hpx~hpx_async_dispatch~hwloc~memkind~numactl~openmp+profil
ing~profiling_load_print~pthread~qthread+serial+shared std=11 ~tests~wrapper arch=linux-ubuntu18.04-broadwell 
<snip>
  >> 13    CMake Error at cmake/kokkos_test_cxx_std.cmake:109 (MESSAGE):                                                         
     14      Invalid compiler for CUDA.  The compiler must be nvcc_wrapper or Clang, but     
```
So, having `kokkos+cuda` does not enable `+wrapper`, and in general, `+wrapper` seems unnecessary. I could be missing something.

Comment (aprokop) [reply var/spack/repos/builtin/packages/kokkos/package.py:168]: Oh, wait, is that because you could compile CUDA with Clang? I retract the point.

Comment (jjwilke) [reply var/spack/repos/builtin/packages/kokkos/package.py:168]: Yep, you got it : )

**Pull request merged**
